Refactoring: typescript

boesing opened the pull request:
<!--
Fill in the relevant information below to help triage your issue.

Pick the target branch based on the following criteria:
  * Documentation improvement: default X.Y.z branch or the oldest support X.Y.z
  * Bugfix: default X.Y.z branch or the oldest support X.Y.z
  * QA improvement (additional tests, CS fixes, etc.) that does not change code
    behavior: default X.Y.z branch or the oldest support X.Y.z
  * New feature, or refactor of existing code: develop branch

You MUST provide a signoff in your commits for us to be able to accept your
patch; you can do this by providing either the --signoff or -s flag when using
"git commit". Please see the project contributing guide and
https://developercertificate.org for details.
-->

|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | no
| QA            | yes

### Description

<!--
Tell us about why this change is necessary:
- Are you fixing a bug or providing a failing unit test to demonstrate a bug?
  - How do you reproduce it?
  - What did you expect to happen?
  - What actually happened?
  - TARGET THE default X.Y.z branch or the oldest support X.Y.z

- Are you adding documentation?
  - TARGET THE default X.Y.z branch or the oldest support X.Y.z

- Are you providing a QA improvement (additional tests, CS fixes, etc.) that
  does not change behavior?
  - Explain why the changes are necessary
  - TARGET THE default X.Y.z branch or the oldest support X.Y.z

- Are you fixing a BC Break?
  - How do you reproduce it?
  - What was the previous behavior?
  - What is the current behavior?
  - TARGET THE default X.Y.z branch or the oldest support X.Y.z

- Are you adding something the library currently does not support?
  - Why should it be added?
  - What will it enable?
  - How will the code be used?
  - TARGET THE develop BRANCH

- Are you refactoring code?
  - Why do you feel the refactor is necessary?
  - What types of refactoring are you doing?
  - TARGET THE develop BRANCH
-->

Converted the application to typescript. 
In combination with #62, this supposed to be a major improvement when it comes to new features.

This also reduces the container size from **426MB** to **188MB** which is a total amount of **~56%**!

#### Features which are implemented in this PR
- #79 
- #82

#### Bugs which are fixed in this PR
- #84 
- #80 


#### Open TODOs
- [x] Implement BC layer for `exclude` which was implemented [here](https://github.com/laminas/laminas-ci-matrix-action/pull/79#discussion_r818499949)
- [x] Add PHP version to the check name
- [x] It seems that the `.laminas-ci.schema.json` is not properly verified, it does not contain `@latest` or `@lowest` for the `php` entry but integration tests are passing

#### Ideas which could be implemented later on
- [x] Generation of `.laminas-ci.schema.json` from typescript interfaces using [typescript-json-schema](https://www.npmjs.com/package/typescript-json-schema) - would target #54 (addressed in https://github.com/laminas/laminas-ci-matrix-action/issues/95)
- [x] Enable code checks or linters when files from tools are changed (i.e. run all PHPUnit checks when `phpunit.xml.dist` is the only file within the diff) (addressed in https://github.com/laminas/laminas-ci-matrix-action/issues/94)
- [x] Extend `exclude` to use the tools `name` along with a `php` and `dependencies` property: `{"name": "PHPUnit", "php": "*"}` (addressed in https://github.com/laminas/laminas-ci-matrix-action/issues/93)
- [x] Optimize `diff` output for better readability in case of differences (addressed in https://github.com/laminas/laminas-ci-matrix-action/issues/92)

boesing committed:
feature: introduce ESLint

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: surround string value when it contains colon

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: provide working github workflow

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: remove `node-version` from `actions/setup-node` to always install the latest version available

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: remove empty-object annotations

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: switch to `npm ci`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: revert weird spacing for `${}` surrounded variables/function calls

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: remove `@babel/eslint-parser`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: use node v17 rather than `current`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
refactoring: convert application to typescript

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: add `tsconfig.json` as `tsconfig.json`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: use script from `package.json` to run linting

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

internalsystemerror commented on .github/workflows/continuous-integration.yml:
```diff
@@ -0,0 +1,22 @@
+  push:
+    branches:
+      - '[0-9]+.[0-9]+.x'
+      - 'refs/pull/*'
```
If you're adding in `package.json`, do you also want to allow renovate patches here?

boesing replied:
Sure, why not.

internalsystemerror replied:
Adding:
```
      - 'renovate/*'
```
should suffice, or this could be added in a future PR dedicated to renovate.

boesing replied:
Yah, I'd love to have a dedicated renovate PR so we can focus on typescript here for now.

internalsystemerror commented on Dockerfile:
```diff
@@ -8,13 +8,18 @@ RUN apk update \
+ADD tsconfig.json /usr/local/source/tsconfig.json
+
+RUN (cd /usr/local/source; npm ci && npm run build)
+RUN mv /usr/local/source/dist/main.js /action/
```
Would it be better to use 2 different layers, one for the build, and the other for the runtime? I haven't thought too much about it, but my instincts are telling me it would be better for caching. Let me know if I'm wrong here.

boesing replied:
TBH: I don't know how. If you have concrete suggestions, feel free to drop them here.

internalsystemerror replied:
It means having 2 `FROM` clauses, the first layer being where the application is compiled, and the second layer just being node with the compiled application copied across (node_modules wouldn't be required). (see https://docs.docker.com/develop/develop-images/multistage-build/#use-multi-stage-builds for more).

@Ocramius Pinging you, as you understand the caching better than I do. Is it worth it? The end result of the layers should be the same, as everything within `/usr/local/source` is removed.

boesing replied:
I reworked the build. Not sure if we already make use of any cache here (guess not) but could be done in a dedicated ticket anyway

internalsystemerror commented on composer.schema.json:
```diff
@@ -3,1286 +3,20 @@
     "title": "Package",
     "type": "object",
     "properties": {
```
This file could be obtained from https://getcomposer.org/schema.json, or for versioning purposes, from https://github.com/composer/composer/blob/2.2/res/composer-schema.json.

boesing replied:
Yah, was already thinking about fetching the schema. But this might be outdated since we do not build the matrix when a new composer version is released.
Same for the container.

internalsystemerror replied:
It being outdated wouldn't be a change from now but this might be one for a future refactor as it works as is.

boesing replied:
Changed the dockerfile to actually receiving the composer schema from `getcomposer.org` while building the container

internalsystemerror commented on package.json:
```diff
@@ -1,8 +1,29 @@
+    "node": "^17"
+  },
+  "type": "commonjs",
+  "devDependencies": {
```
There's no explicit dependency on typescript, and its not semver compliant.

boesing replied:
Ah, what do I have to require?
It actually builds so I havent experienced any problem - but I guess thats because something else is requiring typescript for me?

internalsystemerror replied:
Just add in a dependency to:
```
    "typescript": "~4.6.3"
```
With the `~` constraint instead of `^` since typescript can/does introduce BC breaks in minors.

internalsystemerror commented on src/action/github.ts:
```diff
@@ -0,0 +1,30 @@
+        core.info(message)
+    }
+
+    error(message: string): void {
```
the names of these methods could be more descriptive, e.g. logError. It makes sense for the Logger interface, but when mixed with the action, at first glance one may assume `error` triggers the action to error or `debug` enables debug mode for the action, or something similar.

boesing replied:
I've removed the interface extension and instead added a method `Action#getLogger`. This will actually return the logger so that can be injected and used instead.

internalsystemerror commented on src/app.ts:
```diff
@@ -0,0 +1,175 @@
+            return this.failWithLoggedJsonSchemaValidationErrors(this.composerJsonFileName, schemaValidationResult);
+        }
+
+        schemaValidationResult = this.projectContainsValidJsonConfiguration(
```
Not that it really matters here, but you could use 2 const's over a let which I believe would be a micro-optimisation. (again, correct me if I'm wrong, this could be old news now)

boesing replied:
I am no typescript dev. Not sure if we have to make that kind of micro-optimizations either.
But if we want to do it like this, please make a suggestion as I am happy that it works for the moment 😂

internalsystemerror replied:
If you're happy, I'm happy 😄.

boesing replied:
Well, I am happy with simple things. Not yet a real pro with typescript and thus not too familiar with best practices 😅
Havent spent any time in stuff like this - if you dont mind, I would love to understand the idea behind this so I can improve it and also keep that in mind for future projects 👌🏻

internalsystemerror replied:
~~My assumption is based on the way javascript executes in that it "hoists" variables during the initial creation context where they are first set to be undefined, and then the value calculated whereas consts are defined in a single execution. JS changes often though so I'm not sure how relevant that still is.~~
After a bit of researching, I think this only applies to `var` not `let`.

internalsystemerror commented on src/app.ts:
```diff
@@ -0,0 +1,175 @@
+
+        this.action.debug(`Jobs created: ${JSON.stringify(jobs, null, SPACES_TO_INDENT_JSON)}`);
+
+        return jobs.map((job) => createJobForMatrixFromJob(job)) as [JobForMatrix, ...JobForMatrix[]];
```
Just a question as to `[JobForMatrix, ...JobForMatrix[]]`, wouldn't this be the same as `JobForMatrix[]`?

boesing replied:
non-empty-list

internalsystemerror commented on src/app.ts:
```diff
@@ -0,0 +1,175 @@
+    }
+
+    private projectContainsValidJsonConfiguration(schemaName: PathLike, fileName: PathLike): ValidationResult {
+        const schemaPath = `${this.action.getApplicationDirectory()  }/${  schemaName}`;
```
Spacing inside the curly braces here seems strange.

boesing replied:
❗ 
Good finding, thats from an earlier eslint fix. Have to fix that manually. Will do.

internalsystemerror commented on src/app.ts:
```diff
@@ -0,0 +1,175 @@
+    }
+
+    createConfiguration(filesFromDiff: string[]): Config {
+        return createConfig(
```
Suggestion, `createConfig` to `createConfigObject`.

boesing replied:
❗

internalsystemerror commented on src/app.ts:
```diff
@@ -0,0 +1,175 @@
+
+        if (diff.length === 0) {
+            return {
+                codeChecks : true,
```
If there are no changes, run all checks. But if there are changes, run checks selectively based on what has changed. I assume this is deliberate, as my first thought is that these should be false here.

boesing replied:
selective changes are only made when `diff` is passed.
If there is no `diff` available (due to pushes to a branch for example), we have to execute all checks.

internalsystemerror commented on src/app.ts:
```diff
@@ -0,0 +1,175 @@
+        );
+
+        if (requireCodeChecks) {
+            this.action.info(CODE_CHECKS_ENABLED_MESSAGE);
```
This constant is only used in one place, it could be hard coded. Otherwise, it might be worth moving all messages to a dedicated file (either in an object, or as constants) and including them from there?

boesing replied:
❗ 
Oh yah, it was used more than once before my refactoring. Now its obsolete, true

internalsystemerror commented on src/config/app.ts:
```diff
@@ -0,0 +1,328 @@
+
+function convertJobFromFileToJobs(job: JobFromFile, appConfig: Config): Job[] {
+    const jobDefinition: JobDefinitionFromFile = job.job;
+    const composerDependencies = discoverComposerDependenciesForJob(jobDefinition);
```
`composerDependencies` confused me for a while. Something like `composerRestraints` or `composerDependencyRestraints` may be more descriptive as with dependencies I think, individual required packages.

boesing replied:
Have to google to see what "restraints" is translated to. 😂 

I do not have a meaning here, can change this if we want to rename the initial naming.
I tried to keep as much old stuff as possible since I do not want to totally rework this but first move it to typescript.

internalsystemerror replied:
restraint or constraint, just to imply that this is adding a restriction to the dependency resolution. I wasn't aware the name dependencies was part of the configuration which we don't want to change though, so I'll give way to others opinions here, but when reading through the code, I was expecting this to be a list of required packages.

boesing replied:
Well, as a `constraint` I would assume we are talking about these `~8.0.0` or `^7.4 || ^8.0`, but actually we are talking about `locked`, `latest` and `lowest` dependencies.

boesing replied:
I've renamed it to `dependencySet` where possible.
This might reflect the meaning a little bit better as this will tell composer which dependency set to install:
- `locked` dependencies from `composer.lock`
- `latest` dependencies which are supported by the projects constraints
- `lowest` dependencies which are supported by the projects constraints

internalsystemerror commented on src/config/app.ts:
```diff
@@ -0,0 +1,328 @@
+        return false;
+    }
+
+    return exclusions
```
You could use `some` again here instead of `filter` and the length check.

boesing replied:
What is `some`? Again: I am no typescript/JS dev.

internalsystemerror replied:
```ts
return exclusions
        .filter((exclude) => job.name === exclude.name)
        .length > 0;
```

could be written as:
```ts
return exclusions
        .some((exclude) => job.name === exclude.name);
```

it just returns a boolean if any of the values in the array return true for the callable.

internalsystemerror commented on src/config/app.ts:
```diff
@@ -0,0 +1,328 @@
+            php: phpVersion,
+            phpExtensions: config.phpExtensions,
+            phpIni: config.phpIni,
+            composerDependency: composerDependency,
```
`composerDependency` as with above I think conveys the wrong impression here.

internalsystemerror commented on src/config/app.ts:
```diff
@@ -0,0 +1,328 @@
+
+export function gatherChecks(config: ConfigurationFromFile, appConfig: Config, tools: Tool[]): [Job, ...Job[]] {
+    if (isExplicitChecksConfiguration(config)) {
+        const checks = (config.checks?.map((job) => convertJobFromFileToJobs(job, appConfig)) ?? []).flat(1);
```
Is there a chance for duplicates here? And if so is that desired?

boesing replied:
since these are explicit checks, I don't care about duplicates.
If the project wants to execute the same check multiple times, thats not our problem?

internalsystemerror replied:
Fair

internalsystemerror commented on src/config/app.ts:
```diff
@@ -0,0 +1,328 @@
+    let checks = createChecksForKnownTools(appConfig, tools);
+
+    if (isAdditionalChecksConfiguration(config)) {
+        checks = checks.concat((config.additional_checks?.map((job) => convertJobFromFileToJobs(job, appConfig)) ?? []).flat(1));
```
As with above regarding duplicate array values.

boesing replied:
Again: if the project somehow manages to declare tests multiple times, who cares? I think de-duplication of checks which are partly generated from wildcard stuff and partly hardcoded would be a bigger pain.
If you know a smart solution, we can optimize this after this patch.

As of today (the non-typescript version), there is no deduplication at all and thus I added some in this PR for things like PHP extensions, etc. but these were all just strings.

internalsystemerror replied:
Fair. Happy to give way here. If you do wish to add deduplication, an easy way is to:
```ts
test.filter((value, index, array) => array.indexOf(value) === index);
```

internalsystemerror commented on src/config/input.ts:
```diff
@@ -0,0 +1,79 @@
+}
+
+export interface JobExclusionsFromFile {
+    exclude?: [JobToExcludeFromFile, ...JobToExcludeFromFile[]];
```
`[JobToExcludeFromFile, ...JobToExcludeFromFile[]]` vs `JobToExcludeFromFile[]`?

boesing replied:
non-empty-list

internalsystemerror replied:
Ah yes, of course.

internalsystemerror commented on src/config/input.ts:
```diff
@@ -0,0 +1,79 @@
+}
+
+export interface AdditionalChecksConfigurationFromFile extends ConfigurationFromFile, JobExclusionsFromFile {
+    additional_checks?: [JobFromFile, ...JobFromFile[]];
```
`[JobFromFile, ...JobFromFile[]]` vs `JobFromFile[]`?

boesing replied:
Should contain at least 1 element.
The other annotation also allows no element.

internalsystemerror replied:
Yep, my mistake.

internalsystemerror commented on src/config/input.ts:
```diff
@@ -0,0 +1,79 @@
+}
+
+export interface ExplicitChecksConfigurationFromFile extends ConfigurationFromFile {
+    checks?: [JobFromFile, ...JobFromFile[]];
```
as above, just wondering why `[JobFromFile, ...JobFromFile[]]` vs `JobFromFile[]`?

boesing replied:
non-empty-list

internalsystemerror commented on src/config/input.ts:
```diff
@@ -0,0 +1,79 @@
+    php?: InstallablePhpVersionType | AnyPhpVersionType | LowestPhpVersionType | LatestPhpVersionType;
+    extensions?: string[];
+    ini?: string[];
+    dependencies?: AnyComposerDependency | ComposerDependency;
```
Suggestion, `depedencyVersionRestraint` or `dependencyRestraint` as noted earlier.

boesing replied:
There is no way that we change input config. I do not want to introduce BC breaks just because we are discussing namings.
Input JSON and output JSON is not arguable. We either have to life with it or start deprecating stuff. But the benefit in this case is not really huge, is it?

internalsystemerror replied:
Completely fair, I wasn't thinking about this being used outside of this project.

internalsystemerror commented on src/config/php.ts:
```diff
@@ -0,0 +1,28 @@
+
+export type InstallablePhpVersionType = typeof INSTALLABLE_VERSIONS[number];
+
+export function isInstallableVersion(version: string): version is InstallablePhpVersionType {
```
is the version provided here always in the `x.x` format, with the patch number excluded?

boesing replied:
❗ 
Yes, thats at least how the documentation tells the user to annotate.
We can also enforce this via some kind of JSONSchema `pattern`. 

Marking this as a TODO.

internalsystemerror commented on src/jobs.ts:
```diff
@@ -0,0 +1,21 @@
+import { Tool } from './tools';
+
+export interface JobCreator {
+    createJobs(config: ConfigurationFromFile, appConfig: Config): [Job, ...Job[]];
```
`[Job, ...Job[]]` vs `Job[]`, I'm sure you have a good reason for this and I'll feel very silly once I find out what it is and will probably start using it going forward.

boesing replied:
non-empty-list

internalsystemerror commented on src/main.ts:
```diff
@@ -0,0 +1,50 @@
+}
+
+/* eslint-disable-next-line no-magic-numbers */
+const FIRST_TWO_ARGUMENTS = 2;
```
Why make this a constant? As far as I can see this is only used once (and the value is in the name which suggests this wouldn't be changing). It does make it more descriptive as to what's happening, but that could be accomplished with just a comment.

boesing replied:
Because a constant reflects the meaning of it. When reading `arguments.slice(2)`, thats not as verbose as `arguments.slice(FIRST_TWO_ARGUMENTS)`.

Can remove this, don't have a strong feeling for this tho.

internalsystemerror replied:
I understand, I just would have used a comment if you felt it was confusing.

```
const FIRST_TWO_ARGUMENTS = 2;
const config = app.createConfiguration(process.argv.slice(FIRST_TWO_ARGUMENTS));
```

vs

```
// Create configuration using the first 2 CLI arguments
const config = app.createConfiguration(process.argv.slice(2));
```

These are functionally equivalent though, so happy to go with whatever you prefer.

internalsystemerror commented on src/tools.ts:
```diff
@@ -0,0 +1,131 @@
+}
+
+function detectInfectionCommand(): string {
+    const composerJson: ComposerJson = parseJsonFile('composer.json', true) as ComposerJson;
```
Could this not be reused from elsewhere? I expect hitting the filesystem to be the largest bottleneck in this.

boesing replied:
I don't think that we have to do these kinds of performance improvements and I would prefer not to pass the composer file around the application.

Anyways, thats how it is done right now - we can optimize this somewhere else but for now I would try to focus on the initial idea: improve dev experience when working on our actions.

If we want to do performance improvements, lets do so in dedicated PRs.

internalsystemerror replied:
Fair.

internalsystemerror commented on tsconfig.json:
```diff
@@ -0,0 +1,20 @@
+        "strict": true,
+        "alwaysStrict": true,
+        "esModuleInterop": true,
+        "skipLibCheck": true,
```
Could you also add here:
```
    "noUncheckedIndexedAccess": true,
```
as this isn't included in the `strictNullChecks` which is enabled via `"strict": true"`

boesing replied:
Sure, why not. Will add it with the other fixes.
❗

boesing committed:
qa: add `noUncheckedIndexedAccess`

Since it is not included within `strictNullChecks` which is enabled via the `strict` flag, we want to enable this flag manually.

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: remove unnecessary constant

Due to verbosity, a constant was added for the removal of the first two arguments of the process.
Instead of using that constant, a more verbose comment was added.

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: add `pattern` to `php` JSON schema

By using a pattern, we can verify that the `php` property always contains a version in the format of `\d+.\d+`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
feature: deduplicate the generated additional checks

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
Remove all linting issues

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
Remove spaces added by `eslint --fix`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
Rename `createConfig` to `createConfigObject` to better reflect what is being returned

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
Use `tilde` operator to pin typescript to a patch version

Typescript introduces BC breaks in minors and thus its better to pin it to a specific minor and only allow patches.

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
refactoring: remove `Logger` interface from `Action` interface

Instead, we now do provide a `Action#getLogger` method to provide a logger based on the action we are using.

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: remove redundant constant

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
refactor: rename `ComposerDependency` to `ComposerDependencySet`

To provide a slightly more verbose meaning of what is inside the configuration, renaming `dependency` to `dependencySet` might help.
The dependency set does actually reflect the `lowest` dependencies, `latest` dependencies or `locked` dependencies for the project running this action.

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
feature: split docker build into multi-layer build

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
feature: build container within CI

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: do not login in CI

Due to security regulations, secrets are not accessible in pull requests and thus we should skip the login. Logging in is only required for pushing images and since we do not push, we can skip it safely.

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
feature: download composer JSON schema from getcomposer.org

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: convert file buffer to string before parsing to JSON

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

internalsystemerror reviewed (approved):
I am appeased. This is some good work @boesing.

boesing committed:
bugfix: use enum for `php` property within a `job`

Since we do have to manually add every upcoming PHP version, adding that version to this list might be possible as well.

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

Ocramius commented on .github/workflows/continuous-integration.yml:
```diff
@@ -0,0 +1,38 @@
+      - name: Set up QEMU
+        uses: docker/setup-qemu-action@v1
+
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v1
+
+      - name: Build
+        uses: docker/build-push-action@v2
+        with:
+          push: false
```
Nice! One note: we have no caching on this, so that will probably be a bit slow, but this component changes rarely (urgh, too rarely, sorry!)

boesing replied:
Yah I know but would be fine for me. This is just for CI so we can actually use the final image to run integration tests on and afaik its done only once. 
The dependencies are also very limited and thus its lightning fast.

Ocramius commented on src/action/local.ts:
```diff
@@ -0,0 +1,35 @@
+import {SPACES_TO_INDENT_JSON} from '../json';
+import { Logger } from '../logging';
+
+export class Local implements Action {
```
Don't expect users to run this locally, but is not important.

boesing replied:
This is mainly for development. Without this, its almost impossible to debug stuff. I spent weeks working on this component, I know what I am talking about ;-)

Ocramius commented on src/app.ts:
```diff
@@ -0,0 +1,175 @@
+    }
+
+    private projectContainsValidJsonConfiguration(schemaName: PathLike, fileName: PathLike): ValidationResult {
+        const schemaPath = `${this.action.getApplicationDirectory()}/${schemaName}`;
```
Very convoluted way to concatenate two strings, but okay O_o

internalsystemerror replied:
This is pretty typical in TS

froschdesign replied:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals

boesing replied:
Its also enforced by `tslint`

Ocramius commented on src/config/app.ts:
```diff
@@ -0,0 +1,353 @@
+        return [];
+    }
+
+    return INSTALLABLE_VERSIONS
+        .filter((version) => semver.satisfies(`${version}.0`, composerPhpVersion));
```
Hmm, probably inherent from original design, but this means we need to touch this action at each release 🤔 

Non-issue for porting to TS, just a problem for the future.

boesing replied:
Yes, on each PHP release.
Once.

I don't see any other possibility for projects using `>= 5.6`. How shall we know which PHP versions we do support?
This is also a task because we also have to manually touch the container action to compile that PHP version into the container. So this would be a direct follow-up anywas.

Ocramius commented on src/config/app.ts:
```diff
@@ -0,0 +1,353 @@
+    };
+}
+
+export type IgnorePhpPlatformRequirements = Partial<Record<typeof INSTALLABLE_VERSIONS[number], boolean>>
```
Really cool way to declare a partial :O

Ocramius commented on src/config/composer.ts:
```diff
@@ -0,0 +1,20 @@
+}
+
+/**
+ * @title Package
```
Not needed?

boesing replied:
It was when I created the `composer.schema.json` from this file. But I dropped that idea but did not deleted this annotation. So yes, I will delete it. Thanks!

Ocramius commented on src/json.ts:
```diff
@@ -0,0 +1,33 @@
+
+export const SPACES_TO_INDENT_JSON = 2;
+
+export function parseJsonFile(jsonFile: PathLike, allowMissing: boolean = true): object {
```
eslint doesn't seem to like this: weird.

Still, LGTM.

internalsystemerror replied:
It's because the boolean ia already inferred from the = true, so unless the type could be `boolean|object` for instance, it's saying you may aswell drop it.

Ocramius replied:
If this is normal in typescript (default parameter value restricting parameter type), then no problem :+1:

boesing replied:
I disabled this in the `eslintrc` because heck, I do not want them to "infer" it here. Being more strict is always better IMHO and thus I disabled this at all.

Ocramius commented on src/json.ts:
```diff
@@ -0,0 +1,33 @@
+
+    try {
+        parsedJsonFile = JSON.parse(fs.readFileSync(jsonFile).toString());
+    } catch (error) {
+        let failedReason = JSON.stringify(error);
+
+        if (error instanceof SyntaxError) {
+            failedReason = `${error.message}`;
+        }
+
+        throw new Error(`Failed to parse ${jsonFile}: ${failedReason}`);
+    }
```
Wouldn't it suffice for the json exception to bubble up?

boesing replied:
This file here is re-used and thus, used for multiple JSON files. Thats why I wanted to add the JSON file path to the error here.

Ocramius reviewed (approved):
FML, this has been open for over 6 months? Really sorry for that, @boesing :-( 

I took a slow read through it, my TS-fu is low, but I didn't see any particular scary usages of casts and `any` or `unknown`, so IMO 🚢  at your will.

I will gladly try out the new version on my repos!

boesing committed:
qa: normalize `continuous-integration.yml`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: disable `@typescript-eslint/no-inferrable-types` as we stricter is better

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: disable `object-shorthand` check as we do want to be more explicit

Having "magic" to detect the property name is not really preferred.

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: optimize `eslint` and fixed some issues

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: remove unnecessary annotations

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: remove possibly unused `awesome-typescript-loader`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: tell docker container within GHA that it is actually a GHA

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: pass all host environment variables to the container

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
enhancement: use `GITHUB_ACTIONS` environment variable rather than `GITHUB_ACTION` while also having `Github` as default action

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
docs: add some description on why we do use `GITHUB_ACTIONS` environment variable

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: remove `GITHUB_ACTIONS` environment variable from CI as we do only want to consume that if this action is being used as an action

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: ensure `latest` dependencies for projects without `composer.lock`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: `No checks` check always contains static `locked`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: add integration test for project with `composer.lock`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
revert: change back to `diff` so we get more readable output

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
feature: add PHP version and composer dependency set to output check name

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: restore initial `diff` command

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

internalsystemerror commented on src/app.ts:
```diff
@@ -37,7 +37,7 @@ export class App {
 
         if (!schemaValidationResult.valid) {
-            return this.failWithLoggedJsonSchemaValidationErrors(
+            this.failWithLoggedJsonSchemaValidationErrors(
```
Doesn't this change mean that `sanityChecksPassing()` always returns true?

boesing replied:
Good catch.

I've changed the return type to `void` and handled that in `main.ts`.
Since the `never` methods already handle `exit` on themselves, the `exit` handling within the `main.ts` is not necessary anymore.

boesing replied:
Fixed here: https://github.com/laminas/laminas-ci-matrix-action/pull/83/commits/59ee75d61e8c337fed510701f81175406cae3b58

boesing committed:
qa: add integration tests for `exclude` to verify projects can exclude in the old way

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: rework `sanityChecksPassing` to do `assertions` rather than always returning `true`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

Ocramius commented:
This is starting to look more green-ish 💪

boesing committed:
bugfix: remove unnecessary `node` call from CI as it pollutes the diff

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: remove unnecessary variable `REQUIRE_CHECKS=true` from `entrypoint.sh`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
feature: use `webpack.BannerPlugin` to add shebang to generated `main.js`

This `main.js` is now used in the `entrypoint` without prefixing the executable.
The `main.js` is now also used within the `CI` pipeline to avoid using the `entrypoint.sh` magic which suppresses the `diff` when it is not being detected from the GIT repository

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
feature: output files used for generating the configuration via `debug`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: create test to verify that we do not run code checks for docs only diffs

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: add test for code checks without linting even tho the markdown file does exist

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: add integration test for `.laminas-ci.json` change

This should trigger all checks even tho the diff contains nothing else.

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: ensure that we do not run any check if non-PHP and non-markdown files are changed

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: use more readable length check

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: do not run docker with `-t` since we are not in a TTY

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: show proper step name within job

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: add backwards compatibility for old job exclusions

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
docs: update `exclude` examples to reflect latest changes

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: enhance deprecated exclusion test

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: enhance exclusion test

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: fix tool name in expected matrix output

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: ensure that job definitions are always created via `createJob`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: we should expect `latest` in projects without `composer.lock`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: ensure that expected matrix is generated as expected by adding `composer.json` and `composer.lock`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: initialize variable before pushing

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: `No checks` check will not provide PHP or composer dependency set

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: ensure tests properly reflect what we do want to verify

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: ensure backwards compatibility for deprecated exclusions

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: provide tool information via job to provide backwards compatibility

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: properly detected `@latest` and `@lowest` from job definition for PHP versions

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: ensure line-length is not exceeded

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: we do have `locked` dependencies and thus we should expect `locked` dependencies here

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: wrap ternary operation in brackets

I am not 100% sure what is happening here, but somehow, if the ternary operator is not wrapped in brackets, the null-safe operator does work as expected and the value of the ternary operation is being used instead

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: expected output should not double quote the job

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: since no `composer.lock` is available, we do expect `latest` rather than `locked`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: PHP 8.0 platform requirements are not ignored by default anymore

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: ensure we properly validate the `.laminas-ci.json` with the `laminas-ci.schema.json`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: allow `@lowest` and `@latest` for `php` information within the `.laminas-ci.json`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: use proper `composer.schema.json` name

This also adds an action failure to whenever a schema could not be detected.

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: normalize line endings

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

Ocramius reviewed (approved):
IMO 🚢
